sycamore/transforms/augment_text.py

```diff
@@ -0,0 +1,162 @@
+            except Exception as e:
+                print(f"augment_text: Document {doc.doc_id} failed in sentence {s} with error {e}")
+
+        if len(formatted) == 0:
+            print(
+                f"augment_text: Document {doc.doc_id} ended up with empty text \
+                    representation. Using original text instead"
+            )
+            return doc.text_representation
+        return " ".join(formatted)
+
+
+class UDFTextAugmentor(TextAugmentor):
+    """
+    UDFTextAugmentor augments text by calling a user-defined function (UDF)
+    that maps documents to strings.
+
+    Args:
+        fn (Callable[[Document], str]): A function that maps a document to the
+            string to use as the new `text_representation`
+
+    Example:
+         .. code-block:: python
+
+            def aug_text_fn(doc: Document) -> str:
+                return " ".join([
+                    f"This pertains to the part {doc.properties['part_name']}.",
+                    f"{doc.text_representation}"
+                ])
+            augmentor = UDFTextAugmentor(aug_text_fn)
+            context = sycamore.init()
+            pdf_docset = context.read.binary(paths, binary_format="pdf")
+                .augment_text(augmentor)
+    """
+
+    def __init__(self, fn: Callable[[Document], str]):
+        super().__init__()
+        self._fn = fn
+
+    def augment_text(self, doc: Document) -> Optional[str]:
+        return self._fn(doc)
+
+
+class JinjaTextAugmentor(TextAugmentor):
+    """
+    JinjaTextAugmentor uses a jinja template in a SandboxedEnvironment to
+    transform the text representation with metadata from the thingy
+
+    Args:
+        template (str): A jinja2 template for the new text represenation. Can contain
+            references to `doc` and to any modules passed in the `modules` param
+        modules (dict[str, Any]): A mapping of module names to module objects
+
+    Example:
+         .. code-block:: python
+
+            from sycamore.transforms.augment_text import JinjaTextAugmentor
+            from sycamore.transforms.regex_replace import COALESCE_WHITESPACE
+            import pathlib
+            template = '''This document is from {{ pathlib.Path(doc.properties['path']).name }}.
+            The title is {{ doc.properties['title'] }}.
+            The authors are {{ doc.properties['authors'] }}.
+            {% if doc.text_representation %}
+                {{ doc.text_representation }}
+            {% else %}
+                There is no text representation for this
+            {% endif %}
+            '''
+            aug = JinjaTextAugmentor(template=template, modules={"pathlib": pathlib})
+            aug_docset = exp_docset.augment_text(aug).regex_replace(COALESCE_WHITESPACE)
+            aug_docset.show(show_binary=False, truncate_content=False)
+    """
+
+    def __init__(self, template: str, modules: dict[str, Any] = {}):
+        from jinja2.sandbox import SandboxedEnvironment
```

Comment:
How does the SandboxedEnvironment work? Does it create a separate Python interpreter or something? Just want to understand the implications since it's run in parallel for every task.

Reply:
I find myself liking the Jinja syntax more than I expected, though definitely agree about adding infinite dependencies. @baitsguy since you have actually tried to construct the metadata, what do you think? I suppose if I were to vote I would say keep the UDF version and the Jinja version.

Reply:
As I understand it, the jinja Environment object is essentially an interpreter that runs in python and parses jinja stuff. The SandboxedEnvironment is a subclass of that that can forbid certain operations and stuff. So it's still all in one process, but it's a rather heavyweight function call. That said, it ain't got nothin on openai.

It isn't completely secure, and we might want to add to the security it comes with. But I think it has APIs that make that at least doable. I haven't tried to mess with my filesystem with jinja template rendering, but I'll bet it's not impossible.

Reply:
Got it. As I understand it, this should be instantiated once per worker since you create it in the __init__ call. Is that right? I want to avoid creating a separate sandbox per document, but I don't think this is doing that.

Reply:
yep, should be just one sandbox per worker. Now, we create the same template over and over - jinja doesn't want to pickle templates and that caused problems. Could maybe get around that by doing some sort of check at runtime and if template isn't there then and only then do you compile it, sorta like what the embedder does?

Reply:
What do you mean create the same template? Isn't the template just a string?

Reply:
theres a jinja Template object (`self._env.from_string(source=self._template, globals=self._modules)`). I could store this and just call `Template.render` a whole bunch with different Documents. But it doesn't like to serialize